src/core/mesh/qgsmeshvectorrenderer.cpp

```diff
@@ -518,10 +518,163 @@ QgsMeshVectorRenderer *QgsMeshVectorRenderer::makeVectorRenderer(
         layerExtent,
         datasetMagMaximumValue );
       break;
+    case QgsMeshRendererVectorSettings::WindBarbs:
+      renderer = new QgsMeshVectorWindBarbRenderer(
+        m,
+        datasetVectorValues,
+        datasetValuesMag,
+        datasetMagMaximumValue,
+        datasetMagMinimumValue,
+        dataType,
+        settings,
+        context,
+        size );
+      break;
   }
 
   return renderer;
 }
 
 
+QgsMeshVectorWindBarbRenderer::QgsMeshVectorWindBarbRenderer(
+  const QgsTriangularMesh &m,
+  const QgsMeshDataBlock &datasetValues,
+  const QVector<double> &datasetValuesMag,
+  double datasetMagMaximumValue, double datasetMagMinimumValue,
+  QgsMeshDatasetGroupMetadata::DataType dataType,
+  const QgsMeshRendererVectorSettings &settings,
+  QgsRenderContext &context,
+  QSize size )  : QgsMeshVectorArrowRenderer( m,
+        datasetValues,
+        datasetValuesMag,
+        datasetMagMinimumValue,
+        datasetMagMaximumValue,
+        dataType,
+        settings,
+        context,
+        size )
+{
+}
+
+QgsMeshVectorWindBarbRenderer::~QgsMeshVectorWindBarbRenderer() = default;
+
+void QgsMeshVectorWindBarbRenderer::drawVector( const QgsPointXY &lineStart, double xVal, double yVal, double magnitude )
+{
+  // do not render if magnitude is outside of the filtered range (if filtering is enabled)
+  if ( mCfg.filterMin() >= 0 && magnitude < mCfg.filterMin() )
+    return;
+  if ( mCfg.filterMax() >= 0 && magnitude > mCfg.filterMax() )
+    return;
+
+  QPen pen( mContext.painter()->pen() );
+  pen.setColor( mVectorColoring.color( magnitude ) );
+  mContext.painter()->setPen( pen );
+
+  // we need a brush to fill center circle and pennants
+  QBrush brush( pen.color() );
+  mContext.painter()->setBrush( brush );
+
+  const double shaftLength = mContext.convertToPainterUnits( mCfg.windBarbSettings().shaftLength(),
+                             mCfg.windBarbSettings().shaftLengthUnits() );
+  const double d = shaftLength / 25; // this is a magic number ratio between shaft length and other barb dimensions
+  const double centerRadius = d;
+  const double zeroCircleRadius = 2 * d;
+  const double barbLength = 8 * d + pen.widthF();
+  const double barbAngle = 135;
+  const double barbOffset = 2 * d + pen.widthF();
+
+  // Determine the angle of the vector, counter-clockwise, from east
+  // (and associated trigs)
+  const double vectorAngle = -1.0 * atan( ( -1.0 * yVal ) / xVal ) - mContext.mapToPixel().mapRotation() * M_DEG2RAD;
+  const double cosAlpha = cos( vectorAngle ) * mag( xVal );
+  const double sinAlpha = sin( vectorAngle ) * mag( xVal );
+
+  // Now determine the X and Y distances of the end of the line from the start
+  // Flip the Y axis (pixel vs real-world axis)
+  const double xDist = cosAlpha * shaftLength;
+  const double yDist = - sinAlpha * shaftLength;
+
+  if ( std::abs( xDist ) < 1 && std::abs( yDist ) < 1 )
+    return;
```

Comment:
is this "if" needed? if shaftLength has reasonable size, it should never return...?